### PR TITLE
Updated the PartitionKey and RowKey scheme as discussed in #34

### PR DESCRIPTION
**Pull request by richorama**

### CODE REVIEW

**Comment by jason-bragg** [comment]
If I were told something rolled over every hour, I would intuitively expect it to do so on the hour.  I’m not convinced clipping the reverse timestamp to the hour will line up with clipping now to the hour.  I’d recommend clipping now to the hour, then reversing the timestamp.
Probably a better way, but something like:
var now = DateTime.UtcNow;
var dt = new DateTime(now.Year, now.Month, now.Day, now.Hour, 0, 0, DateTimeKind.Utc);

The reverse order timestamp is generally a useful capability so I'm wondering if these time functions would be better as utility extension functions.
Something like:
var reverseHourlyTimestamp = DateTime.UtcNow.ClipToHour().ReverseTimestamp()

Much of the method seems concerned with the creation of the rowkey.  Does it justify its own function?  Personal preference is for entities to construct their own row key via a static row key formatter.  Having the entity implementation responsible for formatting and parsing row keys (and related values like start/end values for ranged queries) improves encapsulation.
entry.RowKey = StatsTableData.MakeRowKey(name, counter).

**Comment by jason-bragg** [comment]
Looking at the entire function, I noticed a candidate for an opportunistic cleanup, while your in there :).

At the start of the function, we get the statValue, then create the entity, then check the stat value, opting out if there is no value to set.  Might as well check the stat before creating the entity..?

**Comment by gabikliot** [comment]
@richorama 
Can you please provide an example here of what you mean by reverse order. I think I might be a bit confused by that.
Why not just leave the partition key as is, and just change the DATE_FORMAT to include hours.
As for row key, why not just take time.Ticks, or only ticks since the last hour began.
Maybe that is what you are doing now, but it's hard for me to read.

I don't think we want such type of a util function, as it is only used once here.
Same for StatsTableData.MakeRowKey. If it was used twice, I would support a separate function. Otherwise, it hurts readability.

**Comment by veikkoeeva** [comment]
@gabikliot Perhaps these the following explanations are helpful. They look to me they explain in detail what @richorama is doing:

* [Effective way of fetching diagnostics data from Windows Azure Diagnostics Table (Hint: Use PartitionKey)](http://gauravmantri.com/2012/02/17/effective-way-of-fetching-diagnostics-data-from-windows-azure-diagnostics-table-hint-use-partitionkey/) (i.e. how the Azure team is doing this)
* [Using Time-based Partition Keys in #Azure Table Storage](https://alexandrebrisebois.wordpress.com/2014/06/16/using-time-based-partition-keys-in-azure-table-storage/)
* [How to truncate milliseconds off of a .NET DateTime (and other accuracies)](https://stackoverflow.com/questions/1004698/how-to-truncate-milliseconds-off-of-a-net-datetime)

I just noticed this and it's way past my bedtime, but I wonder if the code here could be generalized to extension methods? For instance, this is what I have in my application code currently (F#)

    [<AutoOpen>]
    module Time = 
         open System.Runtime.CompilerServices
    
        [<Extension>]
        type TimeExtensions = 
        
            [<Extension>]
            static member inline Truncate(date: DateTime, resolution) = 
                new DateTime(date.Ticks - (date.Ticks % resolution), date.Kind)
        
            [<Extension>]
            static member inline Truncate(date: DateTimeOffset, resolution) = 
                new DateTime(date.Ticks - (date.Ticks % resolution), DateTimeKind.Utc)

And then

       let DescSortedTableKey(datetime:DateTime) =
            //Logging libraries usually pad with one prefix zero to make the resulting strings even length. The convention is followed here too.
           String.Format("0{0:D19}", DateTime.MaxValue.Ticks - datetime.Ticks)

       let CreateTableRowKey((id:SomeId), (timestamp:DateTime)) =
           let truncationAccuracy = TimeSpan.TicksPerSecond / 2L
           id.ToString() + "_" + timestamp.Truncate(truncationAccuracy).Ticks.ToString()

       let CreatePartitionKey(timestamp:DateTimeOffset) =
           let truncationAccuracy = TimeSpan.TicksPerMinute
           DescSortedTableKey(timestamp.Truncate(truncationAccuracy))

File: src/Orleans/AzureUtils/StatsTableDataManager.cs [hunk 1]
```diff
@@ -147,16 +147,16 @@ private StatsTableData PopulateStatsTableDataEntry(ICounter statsCounter)
             counter++;
             var time = DateTime.UtcNow;
             var timeStr = time.ToString(DATE_FORMAT, CultureInfo.InvariantCulture);
-            entry.PartitionKey = deploymentId + ":" + timeStr;
+
+            // number of ticks remaining until the year 9683
+            var ticks = DateTime.MaxValue.Ticks - time.Ticks;
+
+            // partition the table according to the deployment id and hour
+            entry.PartitionKey = string.Join("$", deploymentId, string.Format("{0:d19}", ticks - ticks % TimeSpan.TicksPerHour));
```
Comment (veikkoeeva) [inline src/Orleans/AzureUtils/StatsTableDataManager.cs]: Should we pad this with a leading zero to make it always be of even length? See also at http://gauravmantri.com/2012/02/17/effective-way-of-fetching-diagnostics-data-from-windows-azure-diagnostics-table-hint-use-partitionkey/ wherein padding is used, albeit one could think with a reference to authority.

File: src/Orleans/AzureUtils/StatsTableDataManager.cs [hunk 2]
```diff
@@ -147,16 +147,16 @@ private StatsTableData PopulateStatsTableDataEntry(ICounter statsCounter)
...
-            {
-                entry.RowKey = name + ":" + counterStr;
-            }
-            else
-            {
-                entry.RowKey = name + ":" + clientEpoch + ":" + counterStr;
-            }
+            
+            // order the rows latest-first in the table 
+            entry.RowKey = string.Join("$", string.Format("{0:d19}", ticks), name, counterStr);
```
Comment (veikkoeeva) [inline src/Orleans/AzureUtils/StatsTableDataManager.cs]: A note here too if padding would make sense here.

**Comment by gabikliot** [comment]
Ok, after reading all the blogs I see now. Thanks for sharing those.
Everyone had good suggestions, its up to you @richorama what you want to change.

**Pull request merged**